staging/src/k8s.io/apiserver/pkg/storage/etcd3/paging_etcd_client.go

```diff
@@ -0,0 +1,110 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package etcd3
+
+import (
+	"context"
+
+	"go.etcd.io/etcd/api/v3/mvccpb"
+	clientv3 "go.etcd.io/etcd/client/v3"
+)
+
+// apiserver only uses the following etcd v3 client options without any sort order and target
+// it default falls back to sorted by keys in lexiographically ascending order
+
+// key
+// WithRange
+// WithLimit
+// WithRev
+
+// etcd mvcc layer increment limit by one to fetch for 'more' flag
+
+// current behavior
+// 1. only set WithLimit when paging is enabled (by default yes) and limit > 0 (it can be 0 if client not set)
+//    1.1 if the resource version is provided, it set up the rev and returnedRV in legacy cases and RVMatchExact
+//    1.2
+// 2. if predicate continue token is provided, restore key and withRev from it
+
+// proposed behavior
+// 1. configure a maximum limit to etcd when limit is not provided by client and construct the full list result
+
+const defaultLimit = 500
+
+type etcdClient struct {
+	c            *clientv3.Client
+	paingEnabled bool
+	limit        int64
+}
+
+// Get implements clientv3.KV.Get
+// APIServer only use key, WithRange, WithLimit and WithRev options without any sort order and target
+// it default falls back to sorted by keys in lexiographically ascending order
+// GetResponse is a type alias of "go.etcd.io/etcd/api/v3/etcdserverpb".RangeResponse
+
+// It's better to add the paging support in APIServer because it knows calling etcd with ascending order and sorting target is key
+func (ec *etcdClient) Get(ctx context.Context, key string, limit int64, getOpts ...clientv3.OpOption) (*clientv3.GetResponse, error) {
+	if !ec.paingEnabled || (limit > 0 && limit <= ec.limit) {
```

Comment:
In the interest of fewer code branches (and thus fewer combinations that need testing), I suggest using MAXINT rather than 0 in the rest of this function - or modify the callers in the same way:
```golang
pageLimit := ec.limit
if !ec.pagingEnabled {
        pageLimit = math.MaxInt
}
if limit == 0 {
        limit = math.MaxInt64
}

// ... and then proceed to the 'paged' version of this function for _all_ cases.
```